src/runtime_src/hip/core/memory_pool.cpp

```diff
@@ -0,0 +1,407 @@
+      dlinkedlist_delete(m_free_list, p0);
+      p1->set_size(p1->get_size() + p0->get_size());
+      p0 = p0->next();
+    }
+  }
+
+  size_t
+  memory_pool_node::free(size_t start)
+  {
+    size_t size_freed = 0;
+    auto used_slot  = m_alloc_list;
+    while (used_slot != nullptr)
+    {
+      if (used_slot->m_start == start )
+      {
+        size_freed = used_slot->get_size();
+        m_used -= size_freed;
+        dlinkedlist_delete(m_alloc_list, used_slot);
+        dlinkedlist_insert(m_free_list, used_slot);
+        used_slot->m_is_free = true;
+        merge_free_slots(used_slot);
+        break;
+      }
+      used_slot = used_slot->m_next;
+    };
+    return size_freed;
+  }
+
+  memory_pool::memory_pool(device* device, size_t max_total_size, size_t pool_size)
+      : m_device(device), m_last_id(0), m_auto_extend(true), m_max_total_size(max_total_size), m_pool_size(pool_size), m_list(), m_mutex(),
+        m_reuse_follow_event_dependencies(1), m_reuse_allow_opportunistic(1), m_reuse_allow_internal_dependencies(1),
+        m_release_threshold(0), m_reserved_mem_current(0), m_reserved_mem_high(0), m_used_mem_current(0), m_used_mem_high(0)
+  {
+    init();
+  }
+
+  void
+  memory_pool::init()
+  {
+    std::lock_guard lock(m_mutex);
+
+    m_reserved_mem_current = m_pool_size;
+
+    if (m_pool_size > m_max_total_size)
+    {
+      throw std::runtime_error("mem poolsize is too big.");
+    }
+    else if (m_pool_size == m_max_total_size)
+    {
+      m_auto_extend = false;
+    }
+
+    m_list.emplace(m_list.end(), std::make_shared<memory_pool_node>(m_device, m_pool_size, m_last_id++));
+  }
+
+  void
+  memory_pool::get_attribute(hipMemPoolAttr attr, void* value)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
+
+    switch (attr)
+    {
+    case hipMemPoolReuseFollowEventDependencies:
+      *reinterpret_cast<int*>(value) = m_reuse_follow_event_dependencies;
+      break;
+
+    case hipMemPoolReuseAllowOpportunistic:
+      *reinterpret_cast<int*>(value) = m_reuse_allow_opportunistic;
+      break;
+
+    case hipMemPoolReuseAllowInternalDependencies:
+      *reinterpret_cast<int*>(value) = m_reuse_allow_internal_dependencies;
+      break;
+
+    case hipMemPoolAttrReleaseThreshold:
+      *reinterpret_cast<uint64_t*>(value) = m_release_threshold;
+      break;
+
+    case hipMemPoolAttrReservedMemCurrent:
+      *reinterpret_cast<uint64_t*>(value) = m_reserved_mem_current;
+      break;
+
+    case hipMemPoolAttrReservedMemHigh:
+      *reinterpret_cast<uint64_t*>(value) = m_reserved_mem_high;
+      break;
+    
+    case hipMemPoolAttrUsedMemCurrent:
+      *reinterpret_cast<uint64_t*>(value) = m_used_mem_current;
+      break;
+
+    case hipMemPoolAttrUsedMemHigh:
+      *reinterpret_cast<uint64_t*>(value) = m_used_mem_high;
+      break;
+    };
+  }
+
+  void
+  memory_pool::set_attribute(hipMemPoolAttr attr, void* value)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
```

Comment:
```suggestion
    if (m_list.empty())
      init();
```

Same pattern repeated many times, so please fix all occurences.